### PR TITLE
Torch-first data loader without dask

### PR DESCRIPTION
**Pull request by jder**
I did some profiling and continue to see a huge amount of time spent in xarray indexing. I built a dataloader which goes as quickly as possible into (CPU) torch and does not use dask. On my laptop this is 3x faster than the eager data loader, and on my test GPU machine it's about 5x faster in the data loader benchmarks. Also seems to eliminate any waiting for data on my GPU machine with 4 workers when running with batch size == 2. 

I manually verified this produces the same data as the eager loader in the debugger -- hoping to add a test for this once @alxmrs is done refactoring the test configuration story a little. 

- [x] add tests for matching old behavior, shape, etc before merging
- [x] should we be masking with NaN?

### CODE REVIEW

File: configs/train_om4.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 
-debug: true
```
Comment (jder) [inline configs/train_om4.yaml:2]: I use this config as "most like real life" so felt like we should remove this.

File: src/ocean_emulators/config.py [hunk 2]
```diff
@@ -41,7 +41,6 @@ class DataConfig:
     num_workers: int = 4
     hist: int = 1
     loader_version: str = str(LoaderVersion.OM4_EAGER.value)
-    use_dask: bool = True
```
Comment (jder) [inline src/ocean_emulators/config.py:44]: We now infer this based on which loader you use.

File: tests/test_datasets.py [hunk 3]
```diff
@@ -455,8 +455,9 @@ def test_profile__loader__1gb(td_loader_pair: LoaderPair, benchmark):
 
     @benchmark
     def bench():
-        for sample in loader:
-            _ = sample
+        indices = np.random.randint(0, len(loader), size=len(loader))
+        for idx in indices:
+            _ = loader.dataset[int(idx)]
```
Comment (jder) [inline tests/test_datasets.py]: Felt like random indexes was more representative

**Comment by alxmrs** [comment]
This is excellent. I really am happy to get this landed. Now that I've written an Xarray-based loader, I have a better understanding of what's going on. I think two additional improvements we could adapt from the lazy loader are: 
- Making the data up front once (not on the `__getitem__` loop)
- Vectorizing steps

**Comment by jder** [comment]
Agreed about the vectorizing, it should not be hard. I'm not sure we can do the masking up front, though, without loading all data into ram.

File: src/ocean_emulators/datasets.py [hunk 4]
```diff
@@ -594,3 +594,172 @@ def _get_label(self, x_index) -> Prognostic:
...
+
+        self.hist: int = hist
+        self.steps: int = steps
+        self.stride: int = stride
+
+        self.num_prognostic_channels: int = (hist + 1) * len(prognostic_var_names)
+        self._prognostic_vars: xr.Dataset = data[prognostic_var_names]
+        self._boundary_vars: xr.Dataset = data[boundary_var_names]
+
+        # This class will be used only for training
```
Comment (alxmrs) [inline src/ocean_emulators/datasets.py]:
```suggestion
        # This class will be used only for training and validation
```

File: src/ocean_emulators/datasets.py [hunk 5]
```diff
@@ -594,3 +594,172 @@ def _get_label(self, x_index) -> Prognostic:
...
+        return total_input, label
+
+    def _prep_prognostic_steps(
+        self, prognostic_steps: Float[torch.Tensor, "variable time lat lon"]
+    ) -> Input:
+        normalize = Normalize.get_instance()
+        prognostic_steps = rearrange(
+            prognostic_steps,
+            "variable time lat lon -> time variable lat lon",
+        )
```
Comment (alxmrs) [inline src/ocean_emulators/datasets.py:734]: 🐑  Is a named `transpose` available? it may be slightly simpler.

File: src/ocean_emulators/datasets.py [hunk 6]
```diff
@@ -594,3 +594,172 @@ def _get_label(self, x_index) -> Prognostic:
...
+        ).float()
+
+        # flatten time and variable dimensions into a set of channels for model
+        prognostic_steps = rearrange(
+            prognostic_steps,
+            "time variable lat lon -> (time variable) lat lon",
+        )
+
+        # post-normalize, mask out values where there is no ocean
+        prognostic_steps = torch.where(self.wet, prognostic_steps, 0.0)
```
Comment (alxmrs) [inline src/ocean_emulators/datasets.py:748]: Future PR: we should consider moving this to the `__init__`.

File: tests/test_datasets.py (hunk 3 above)
Comment (alxmrs) [reply tests/test_datasets.py]: SGTM

File: src/ocean_emulators/datasets.py [hunk 7]
```diff
@@ -594,3 +594,172 @@ def _get_label(self, x_index) -> Prognostic:
...
+
+        return prognostic_steps
+
+    def _get_x_index(
+        self, idx: int, step: int, prev_rolling_idx: int | None
```
Comment (alxmrs) [inline src/ocean_emulators/datasets.py]: 🐑  I know you're just copying over the previous version, but maybe we can get rid of `prev_rolling_idx`?

File: src/ocean_emulators/datasets.py (hunk 5 above)
Comment (alxmrs) [reply src/ocean_emulators/datasets.py:734]: Is this needed for normalize?

File: src/ocean_emulators/datasets.py [hunk 8]
```diff
@@ -594,3 +594,172 @@ def _get_label(self, x_index) -> Prognostic:
...
+            input_, label = self._get_input_and_label(prognostic_all, boundary)
+
+            TD.insert(
+                input_=input_,
+                label=label,
+            )
+
+        return TD
+
+    def _get_input_and_label(
```
Comment (alxmrs) [inline src/ocean_emulators/datasets.py:707]: I like how you combine all prognostic access into a single function.

File: src/ocean_emulators/datasets.py (hunk 5 above)
Comment (alxmrs) [reply src/ocean_emulators/datasets.py:734]: Could this be done in the caller function?

File: src/ocean_emulators/datasets.py [hunk 9]
```diff
@@ -594,3 +594,172 @@ def _get_label(self, x_index) -> Prognostic:
...
+        self,
+        # time includes (self.hist + 1) past steps and the (label) future steps
+        prognostic_all: Float[torch.Tensor, "variable time lat lon"],
+        boundary: Float[torch.Tensor, "variable lat lon"],
+    ) -> tuple[Input, Prognostic]:
+        normalize = Normalize.get_instance()
```
Comment (suryadheeshjith) [inline src/ocean_emulators/datasets.py:713]: Does this have to be here? Are you getting errors when you get the instance in init?

File: src/ocean_emulators/datasets.py (hunk 6 above)
Comment (suryadheeshjith) [reply src/ocean_emulators/datasets.py:748]: I think the entire loader here is created so dask is not being used and this would mean normalize would have to occur during loading since we cannot normalize the entire dataset at once

Comment (alxmrs) [reply src/ocean_emulators/datasets.py:748]: Same with the normalization

**Review by suryadheeshjith — APPROVED** [review]
LGTM overall

File: src/ocean_emulators/train.py [hunk 10]
```diff
@@ -157,7 +158,8 @@ def __init__(self, cfg: TrainConfig) -> None:
         self.data_stds_path = cfg.data.data_stds_path
         self.scaling_residuals_file = cfg.data.scaling_residuals_file
 
-        if cfg.data.use_dask:
+        self.use_dask = cfg.data.loader_version != LoaderVersion.OM4_TORCH.value
```
Comment (alxmrs) [inline src/ocean_emulators/train.py:163]: 🐑  maybe this could be made overridable by the config

File: src/ocean_emulators/datasets.py (hunk 6 above)
Comment (suryadheeshjith) [reply src/ocean_emulators/datasets.py:748]: Oh I thought you were referencing the normalization code

File: src/ocean_emulators/utils/data.py [hunk 11]
```diff
@@ -391,3 +391,22 @@ def unnormalize_tensor_prognostic(
...
+        if data.ndim == 3:
+            tensor_mean = tensor_mean.reshape([-1, 1, 1])
+            tensor_std = tensor_std.reshape([-1, 1, 1])
+        else:
+            raise ValueError(f"Invalid data shape: {data.shape}")
+
+        norm = (data - tensor_mean) / tensor_std
+        if fill_nan:
+            norm = norm.nan_to_num(nan=fill_value)
+        norm = norm.to(data.dtype)
```
Comment (alxmrs) [inline src/ocean_emulators/utils/data.py:411]: may not be necessary since we cast to `float` right after in the calling function.

File: src/ocean_emulators/datasets.py (hunk 6 above)
Comment (alxmrs) [reply src/ocean_emulators/datasets.py:748]: Sorry Surya and Jesse, I missed both your comments. Ah, I now understand that this saves us from loading all the data into memory. 

In that case, I wonder if we could push it to the data engineering. Short of that, I think this is fine.

**Review by alxmrs — APPROVED** [review]
Approved, will review again after I merge my testing PR.

File: src/ocean_emulators/datasets.py (hunk 9 above)
Comment (jder) [reply src/ocean_emulators/datasets.py:713]: I removed the fetch of the instance in init; generally I try to avoid having state when I don't need it but it would work either way.

File: src/ocean_emulators/datasets.py (hunk 5 above)
Comment (jder) [reply src/ocean_emulators/datasets.py:734]: I'm not sure what you mean by "a named transpose" -- this is a torch vector at this point so it doesn't know what axes are what. And yes, it's required for the next step as the comment says. 

It could be done by the caller but then it would be duplicated for input + label. Why would you prefer the caller?

File: src/ocean_emulators/datasets.py (hunk 6 above)
Comment (jder) [reply src/ocean_emulators/datasets.py:748]: Yeah, agreed that doing this in the data someday would be nice.

File: src/ocean_emulators/utils/data.py (hunk 11 above)
Comment (jder) [reply src/ocean_emulators/utils/data.py:411]: This matches all the other normalize functions in this file so would rather leave it. (The caller's float will be a no-op if so, right?)

File: src/ocean_emulators/train.py (hunk 10 above)
Comment (jder) [reply src/ocean_emulators/train.py:163]: I think using dask with the torch loader will be slower and not using dask with the other loaders will load all data into memory, so I'm not sure why we would want to configure it separately.

File: src/ocean_emulators/datasets.py (hunk 9 above)
Comment (suryadheeshjith) [reply src/ocean_emulators/datasets.py:713]: Alright

**Comment by jder** [comment]
@alxmrs can I get you to take a quick look at the [new tests](https://github.com/suryadheeshjith/Ocean_Emulator/pull/178/commits/22514eeaea0e0e51f96d015c681d33422f9af7ce) before I merge?

File: src/ocean_emulators/datasets.py (hunk 5 above)
Comment (alxmrs) [reply src/ocean_emulators/datasets.py:734]: Happy to have it here, thank for pointing out things that I missed.

File: tests/test_datasets.py [hunk 12]
```diff
@@ -347,10 +371,17 @@ def test_inference__data_is_not_zero(inference_loader_pair):
             )
 
 
-def test_om4__is_equal_to_v1_data_loader(train_config):
+ORIGINAL_LOADER_VERSION = LoaderVersion.OM4_EAGER
```
Comment (alxmrs) [inline tests/test_datasets.py:374]: This LGTM, but can we move it to the top of the file?

File: tests/test_datasets.py [hunk 13]
```diff
@@ -347,10 +371,17 @@ def test_inference__data_is_not_zero(inference_loader_pair):
...
+
+
+@pytest.mark.parametrize(
+    "loader_version", [v for v in LoaderVersion if v != ORIGINAL_LOADER_VERSION]
+)
```
Comment (alxmrs) [inline tests/test_datasets.py:379]: Nice, good idea.

**Review by alxmrs — APPROVED** [review]
LGTM

File: tests/test_datasets.py (hunk 12 above)
Comment (jder) [reply tests/test_datasets.py:374]: Why? It's only meaningful in the context of this test. If you don't want it floating here, maybe I make a class for it + the test to live in?

Comment (alxmrs) [reply tests/test_datasets.py:374]: I don't have a good reason except to say that I'm used to seeing constants at the top of a file. It's more of a 🐑 ish suggestion; I'm happy to leave it here.
